src/jit/smallhash.h

```diff
@@ -0,0 +1,586 @@
+//    TKeyInfo - A type that conforms to the HashTableInfo<TKey> concept.
+template<typename TKey, typename TValue, typename TKeyInfo = HashTableInfo<TKey>>
+class HashTableBase
+{
+    friend class KeyValuePair;
+    friend class Iterator;
+
+protected:
+    //------------------------------------------------------------------------
+    // HashTableBase::Bucket: provides storage for the key-value pairs that
+    //                        make up the contents of the table.
+    //
+    // The "home" bucket for a particular key is the bucket indexed by the
+    // key's hash code modulo the size of the bucket array (the "home index").
+    //
+    // The home bucket is always considered to be part of the chain that it
+    // roots, even if it is also part of the chain rooted at a different
+    // bucket. `m_firstOffset` indicates the offset of the first non-home
+    // bucket in the home bucket's chain. If the `m_firstOffset` of a bucket
+    // is 0, the chain rooted at that bucket is empty.
+    //
+    // The index of the next bucket in a chain is calculated by adding the
+    // value in `m_nextOffset` to the index of the current bucket. If
+    // `m_nextOffset` is 0, the current bucket is the end of its chain. Each
+    // bucket in a chain must be occupied (i.e. `m_isFull` will be true).
+    struct Bucket
+    {
+        bool m_isFull;          // True if the bucket is occupied; false otherwise.
+
+        unsigned m_firstOffset; // The offset to the first node in the chain for this bucket index.
+        unsigned m_nextOffset;  // The offset to the next node in the chain for this bucket index.
+
+        unsigned m_hash;        // The hash code for the element stored in this bucket.
+        TKey m_key;             // The key for the element stored in this bucket.
+        TValue m_value;         // The value for the element stored in this bucket.
+    };
+
+private:
+    Compiler* m_compiler;      // The compiler context to use for allocations.
+    Bucket* m_buckets;         // The bucket array.
+    unsigned m_numBuckets;     // The number of buckets in the bucket array.
+    unsigned m_numFullBuckets; // The number of occupied buckets.
+
+    //------------------------------------------------------------------------
+    // HashTableBase::Insert: inserts a key-value pair into a bucket array.
+    //
+    // Arguments:
+    //    buckets    - The bucket array in which to insert the key-value pair.
+    //    numBuckets - The number of buckets in the bucket array.
+    //    hash       - The hash code of the key to insert.
+    //    key        - The key to insert.
+    //    value      - The value to insert.
+    //
+    // Returns:
+    //    True if the key-value pair was successfully inserted; false
+    //    otherwise.
+    static bool Insert(Bucket* buckets, unsigned numBuckets, unsigned hash, const TKey& key, const TValue& value)
+    {
+        const unsigned mask = numBuckets - 1;
+        unsigned homeIndex = hash & mask;
```

Comment:
This works correctly only the number of buckets is a power of 2. Because of this the initial hash table size should also be constrained to be a power of 2.